### PR TITLE
Discrete Gramians

### PR DESCRIPTION
**Pull request by artpelling**
Related to #1522, this PR adds
- adds bindings for `slycot` and `scipy` solvers for discrete Lyapunov equations
- adds functionality to compute gramians (and norms and hsvs) for discrete-time LTI-systems
- as a consequence of the discussion in #1522, the solvers for continuous and discrete time are implemented in designated functions. the current ones are renamed, e.g. `solve_cont_lyap_lrcf`. the new ones are called `solve_disc_lyap_lrcf` etc. In a previous version, the solvers could be switched with a `cont_time` argument which made the handling of the backends messy.
- ~new default backend variables are introduced in lyapunov.py~ backend default variables are unified in a `FrozenDict` of `FrozenDicts`
- adds `NotImplemented` errors in some reductors in `bt.py`

It contains a workaround 2e0eeef9161dda7e198431fbb5c772b5b9e7f843 for an incorrectly set `ldwork` variable in `slycot` which can hopefully be removed again in the future.

This PR should be only merged after #1500.

Todos:
- [x] Merge #1500
- [x] write tests
- [x] update docstrings

### CODE REVIEW

File: src/pymor/algorithms/lyapunov.py
Comment (artpelling) [inline src/pymor/algorithms/lyapunov.py:12]: is the number of defaults with very long names getting out of hand?

File: src/pymor/bindings/slycot.py
Comment (artpelling) [inline src/pymor/bindings/slycot.py:180]: I have created a PR for this already https://github.com/python-control/Slycot/pull/166

File: src/pymor/algorithms/lyapunov.py
Comment (artpelling) [inline src/pymor/algorithms/lyapunov.py:26]: is the number of defaults with very long names getting out of hand?

File: src/pymor/bindings/slycot.py
Comment (artpelling) [reply src/pymor/bindings/slycot.py:180]: It has been merged into Slycot/master and should appear in the next release.

**Comment by artpelling** [comment]
I have added some tests. There must be a mistake somewhere, because `test_disc_lrcf` fails. This is weird to me, because `test_disc_dense` does not and the tests and called functions have exactly the same structure as the coninuous case.

File: src/pymor/models/iosys.py
Comment (pmli) [inline src/pymor/models/iosys.py:585]:
```suggestion
        solve_lyap_lrcf = solve_cont_lyap_lrcf if self.sampling_time == 0 else solve_disc_lyap_lrcf
```

Similar below.

**Review by pmli — COMMENTED** [review]
I added some comments.

I'm still thinking about whether it is good to have separate `solve_cont_lyap` and `solve_disc_lyap`. It looks more or less ok so far, but there is a significant amount of repeated code.

Regarding tests, it might be good to use stable matrices in the discrete-time sense for discrete-time Lyapunov equation solvers.

**Comment by artpelling** [comment]
> I'm still thinking about whether it is good to have separate `solve_cont_lyap` and `solve_disc_lyap`. It looks more or less ok so far, but there is a significant amount of repeated code.

Me, too. I have split it into two functions after the discussion in #1522. I will attend your comments next week. Ideally after we can reach a final agreement on this topic.

File: src/pymor/algorithms/lyapunov.py
Comment (artpelling) [reply src/pymor/algorithms/lyapunov.py]: I have actually omitted the 'lrcf' option, because I could not think of a name for "the opposite of lrcf" and that option was identical to non-lrcf dense/sparse anyway.

File: src/pymor/bindings/slycot.py
Comment (artpelling) [reply src/pymor/bindings/slycot.py]: I have made a compromise now. Lyapunov has dedicated functions, the backends only one with a `cont_time` argument.

**Comment by artpelling** [comment]
> Regarding tests, it might be good to use stable matrices in the discrete-time sense for discrete-time Lyapunov equation solvers.

I have added a simple step that scales down A accordingly, I hope that is OK.

File: src/pymortests/lyapunov.py
Comment (pmli) [inline src/pymortests/lyapunov.py:169]: Since the matrices come from a semi-discretized PDE, I guess that a more natural way to generate Schur-stable matrices would be to use a fully discretized PDE, e.g., using explicit Euler time-stepping. In `conv_diff_1d_*` functions, you could add a `cont_time` argument, and if it is `False`, use `A = sps.eye(n) + dt * A` or `A = E + dt * A` (`dt = 0.1 / (4*a*(n+1)**2 + b*(n+1))` should work).

**Review by pmli — COMMENTED** [review]
I see that you used `solve_lyap_lrcf` and `solve_lyap_dense` with a `cont_time` argument for the backends. I like that it reduces code duplication, but would like to hear what others think.

For the interface in `algorithms/lyapunov.py`, there are separate methods for continuous and discrete-time cases. I guess the reason is that otherwise, the docstrings would be even more huge? It is inconsistent with backend solvers, but it might be a good compromise.

**Comment by artpelling** [comment]
Thanks for your reviews and sorry about the messy docstrings, looks like I got lost in some cases, because there was so much to change. I have fixed everything that you suggested.

Let me know, if you agree with https://github.com/pymor/pymor/pull/1525/commits/cb02c7fb0810b9991cefe1256a6717a3e27a31e1

> I see that you used `solve_lyap_lrcf` and `solve_lyap_dense` with a `cont_time` argument for the backends. I like that it reduces code duplication, but would like to hear what others think.
> For the interface in `algorithms/lyapunov.py`, there are separate methods for continuous and discrete-time cases. I guess the reason is that otherwise, the docstrings would be even more huge? It is inconsistent with backend solvers, but it might be a good compromise.

Exactly. I think this one is the most sensible. Having now gone through three variants, it would be nice to have everyone on board before changes are made again :smiling_face_with_tear:

File: src/pymor/bindings/scipy.py
Comment (pmli) [inline src/pymor/bindings/scipy.py:426]:
```suggestion
        if cont_time:
            X = solve_continuous_lyapunov(A, -B @ B.T)
        else:
            X = solve_discrete_lyapunov(A, B @ B.T)
```

**Review by pmli — COMMENTED** [review]
Just some minor comments.
Also, I noticed that the docstring of `LTIModel.gramian` should be updated (concerning the system assumptions).

**Review by pmli — APPROVED** [review]

File: src/pymor/algorithms/lyapunov.py
Comment (sdrave) [inline src/pymor/algorithms/lyapunov.py:14]: What is your reason for going with `FrozenDict` here? I'd say it's fine to use a plain dict here (and leave some room for hacking). So far, `FrozenDict` is only used to ensure immutability of immutable objects, where it is relatively likely that  changing values could cause unexpected side effects

Comment (artpelling) [reply src/pymor/algorithms/lyapunov.py:14]: @pmli suggested this. I think the idea is that these are built-in defaults and different values can always be passed to the functions. I'm happy to change it though :)

Comment (pmli) [reply src/pymor/algorithms/lyapunov.py:14]: @sdrave The solver backend could already be modified via `defaults` of `solve_cont_lyap_lrcf` and other methods. Having a mutable global variable feels a bit strange to me...

Comment (sdrave) [reply src/pymor/algorithms/lyapunov.py:14]: I know. I still do not see any real benefit in using `FrozenDict` here so I'd leave it out. But this is probably a matter of taste, so I would leave it up to @artpelling's taste ..

Comment (artpelling) [reply src/pymor/algorithms/lyapunov.py:14]: Don't make me choose sides :cold_sweat:

To resolve this, I would stay with the `FrozenDict` as no functionality is lost and there does not yet seem to be a unified way to handle this. If we decide on a paradigm globally this should be easily adjusted.

Comment (sdrave) [reply src/pymor/algorithms/lyapunov.py:14]: Fine with me ..

**Review by sdrave — APPROVED** [review]

**Comment by pmli** [comment]
@lbalicki What do you think?

File: src/pymor/reductors/bt.py
Comment (pmli) [inline src/pymor/reductors/bt.py:140]: Why would the standard BT not work?

Comment (artpelling) [reply src/pymor/reductors/bt.py:140]: I thought the same but I got an error when testing because `LTIPGReductor.build_rom` in line 340 does not set `sampling_time`. I would like to attend to the BT methods in a dedicated PR once this is merged.

Comment (pmli) [reply src/pymor/reductors/bt.py:140]: Ok, sounds good.

**Review by lbalicki — APPROVED** [review]
Sorry for the late response. The PR looks good to me and I'm fine with the names of the `solve_lyap` functions as they currently are in the PR.

**Comment by pmli** [comment]
Thanks @artpelling.

**Pull request merged**
